# ETL Project

Hi again! For this project, one was asked to extract information from at least 3 different sources, using a minimum of 2 different methods and, finally, load a table to SQL

For this project I'll choose:
+ 1st Pokémon generation (150) and Mew and the 151


+ Information sources:
    + Kaggle: extract and csv from there
    + Pokémon API
    + Bulbapedia


+ Methods:
    + CSV upload
    + API
    + Web Scrapping (using Selenium)

# CSV extracting

Firstly, let us extract and read a csv file from Kaggle. I've done this already, and you can find the same file in the following link: https://www.kaggle.com/alopez247/pokemon

In [1]:
import numpy as np

In [2]:
import pandas as pd

df = pd.read_csv('pokemon_alopez247.csv')

In [3]:
df.tail(5)

Number       Name   Type_1  Type_2  Total   HP  Attack  Defense  Sp_Atk  \
716     717    Yveltal     Dark  Flying    680  126     131       95     131   
717     718    Zygarde   Dragon  Ground    600  108     100      121      81   
718     719    Diancie     Rock   Fairy    600   50     100      150     100   
719     720      Hoopa  Psychic   Ghost    600   80     110       60     150   
720     721  Volcanion     Fire   Water    600   80     110      120     130   

     Sp_Def  ...   Color  hasGender  Pr_Male   Egg_Group_1  Egg_Group_2  \
716      98  ...     Red      False      NaN  Undiscovered          NaN   
717      95  ...   Green      False      NaN  Undiscovered          NaN   
718     150  ...    Pink      False      NaN  Undiscovered          NaN   
719     130  ...  Purple      False      NaN  Undiscovered          NaN   
720      90  ...   Brown      False      NaN  Undiscovered          NaN   

     hasMegaEvolution Height_m Weight_kg  Catch_Rate       Body_Style  
716             False     5.79     203.0          45        two_wings  
717             False     5.00     305.0           3  serpentine_body  
718              True     0.71       8.8           3        head_arms  
719             False     0.51       9.0           3        head_only  
720             False     1.70     195.0           3        quadruped  

[5 rows x 23 columns]

I don't need so many Pokémons... I know now the first generation! Let us remove any Pokémon that doesn't belong to this generation, ie, from the #152 Pokémon forward:

In [4]:
df = df.drop(range(151,721,1))

In [5]:
df

Number        Name   Type_1  Type_2  Total   HP  Attack  Defense  Sp_Atk  \
0         1   Bulbasaur    Grass  Poison    318   45      49       49      65   
1         2     Ivysaur    Grass  Poison    405   60      62       63      80   
2         3    Venusaur    Grass  Poison    525   80      82       83     100   
3         4  Charmander     Fire     NaN    309   39      52       43      60   
4         5  Charmeleon     Fire     NaN    405   58      64       58      80   
..      ...         ...      ...     ...    ...  ...     ...      ...     ...   
146     147     Dratini   Dragon     NaN    300   41      64       45      50   
147     148   Dragonair   Dragon     NaN    420   61      84       65      70   
148     149   Dragonite   Dragon  Flying    600   91     134       95     100   
149     150      Mewtwo  Psychic     NaN    680  106     110       90     154   
150     151         Mew  Psychic     NaN    600  100     100      100     100   

     Sp_Def  ...   Color  hasGender  Pr_Male   Egg_Group_1  Egg_Group_2  \
0        65  ...   Green       True    0.875       Monster        Grass   
1        80  ...   Green       True    0.875       Monster        Grass   
2       100  ...   Green       True    0.875       Monster        Grass   
3        50  ...     Red       True    0.875       Monster       Dragon   
4        65  ...     Red       True    0.875       Monster       Dragon   
..      ...  ...     ...        ...      ...           ...          ...   
146      50  ...    Blue       True    0.500       Water_1       Dragon   
147      70  ...    Blue       True    0.500       Water_1       Dragon   
148     100  ...   Brown       True    0.500       Water_1       Dragon   
149      90  ...  Purple      False      NaN  Undiscovered          NaN   
150     100  ...    Pink      False      NaN  Undiscovered          NaN   

     hasMegaEvolution Height_m Weight_kg  Catch_Rate       Body_Style  
0               False     0.71       6.9          45        quadruped  
1               False     0.99      13.0          45        quadruped  
2                True     2.01     100.0          45        quadruped  
3               False     0.61       8.5          45   bipedal_tailed  
4               False     1.09      19.0          45   bipedal_tailed  
..                ...      ...       ...         ...              ...  
146             False     1.80       3.3          45  serpentine_body  
147             False     3.99      16.5          45  serpentine_body  
148             False     2.21     210.0          45   bipedal_tailed  
149              True     2.01     122.0           3   bipedal_tailed  
150             False     0.41       4.0          45   bipedal_tailed  

[151 rows x 23 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 151 entries, 0 to 150
Data columns (total 23 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Number            151 non-null    int64  
 1   Name              151 non-null    object 
 2   Type_1            151 non-null    object 
 3   Type_2            67 non-null     object 
 4   Total             151 non-null    int64  
 5   HP                151 non-null    int64  
 6   Attack            151 non-null    int64  
 7   Defense           151 non-null    int64  
 8   Sp_Atk            151 non-null    int64  
 9   Sp_Def            151 non-null    int64  
 10  Speed             151 non-null    int64  
 11  Generation        151 non-null    int64  
 12  isLegendary       151 non-null    bool   
 13  Color             151 non-null    object 
 14  hasGender         151 non-null    bool   
 15  Pr_Male           138 non-null    float64
 16  Egg_Group_1       151 non-null    object 
 1

I'll do the following changes of the following data set:

+ Remove the columns:
    + Generation: I already know they are all from the 1st generation (Mew might not be, but let us consider he is)
    + Pr_Male: probability of being male. Not interesting from my point of view (although, I'll leave the column hasGender for curiosity matters:
    
        df.iloc[:,15]
        
        0      0.875
        
        1      0.875
            ...
        
        146    0.500
        
        147    0.500
        
        Name: Pr_Male, Length: 151, dtype: float64
  
    + Egg_Group_1
    + Egg_Group_2
    + hasMegaEvolution

In [7]:
df.drop(['Generation', 'Pr_Male', 'Egg_Group_1', 'Egg_Group_2', 'hasMegaEvolution'], axis = 1, inplace = True)

In [8]:
df.shape #We now have 18 columns!

(151, 18)

Luckly for us, the columns from our Dataframe are very clean. Let us just lower case them!

In [9]:
new_columns = [c.lower() for c in df.columns]
df=df.rename(columns={k:v for k,v in zip(df.columns, new_columns)})
df.columns

Index(['number', 'name', 'type_1', 'type_2', 'total', 'hp', 'attack',
       'defense', 'sp_atk', 'sp_def', 'speed', 'islegendary', 'color',
       'hasgender', 'height_m', 'weight_kg', 'catch_rate', 'body_style'],
      dtype='object')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 151 entries, 0 to 150
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   number       151 non-null    int64  
 1   name         151 non-null    object 
 2   type_1       151 non-null    object 
 3   type_2       67 non-null     object 
 4   total        151 non-null    int64  
 5   hp           151 non-null    int64  
 6   attack       151 non-null    int64  
 7   defense      151 non-null    int64  
 8   sp_atk       151 non-null    int64  
 9   sp_def       151 non-null    int64  
 10  speed        151 non-null    int64  
 11  islegendary  151 non-null    bool   
 12  color        151 non-null    object 
 13  hasgender    151 non-null    bool   
 14  height_m     151 non-null    float64
 15  weight_kg    151 non-null    float64
 16  catch_rate   151 non-null    int64  
 17  body_style   151 non-null    object 
dtypes: bool(2), float64(2), int64(9), object(5)
memory

Inspecting the above information, one could consider replacing the NaN values from the column type_2. It happens that not all Pokémon have 2 types. Therefore, I'll leave as it is for now!

# API Extracting

For this one, we should take import the following libraries:

In [11]:
!pip install requests
import json
import requests
from pandas.io.json import json_normalize

You should consider upgrading via the 'c:\users\ffrei\appdata\local\programs\python\python39\python.exe -m pip install --upgrade pip' command.


We are going go collecto some qualitative information regarding the Pokémon. For that, we'll use the following API:

https://pokeapi.co/api/v2/pokemon-species/{id or name}/, being id the number of the Pokémon or its name.

Let's do an example with one Pokémon, and expand it to the 151

In [12]:
BASE_URL = 'https://pokeapi.co/api/v2/pokemon-species/150/'
results = requests.get(BASE_URL).json()
results.keys()

dict_keys(['base_happiness', 'capture_rate', 'color', 'egg_groups', 'evolution_chain', 'evolves_from_species', 'flavor_text_entries', 'form_descriptions', 'forms_switchable', 'gender_rate', 'genera', 'generation', 'growth_rate', 'habitat', 'has_gender_differences', 'hatch_counter', 'id', 'is_baby', 'is_legendary', 'is_mythical', 'name', 'names', 'order', 'pal_park_encounters', 'pokedex_numbers', 'shape', 'varieties'])

In [13]:
results['flavor_text_entries'][0]['flavor_text']

'It was created by\na scientist after\nyears of horrific\x0cgene splicing and\nDNA engineering\nexperiments.'

So, we want to collect the information contained in the key 'flavor_text_entries' which corresponds to the value 'flavor_text'. We should proceed as follow:

In [14]:
BASE_URL = 'https://pokeapi.co/api/v2/pokemon-species/'
links = []

for i in range(152):
    if i != 0:
        result = requests.get(BASE_URL + str(i) + '/').json()
        links.append(result)

In [15]:
links[0].keys()

dict_keys(['base_happiness', 'capture_rate', 'color', 'egg_groups', 'evolution_chain', 'evolves_from_species', 'flavor_text_entries', 'form_descriptions', 'forms_switchable', 'gender_rate', 'genera', 'generation', 'growth_rate', 'habitat', 'has_gender_differences', 'hatch_counter', 'id', 'is_baby', 'is_legendary', 'is_mythical', 'name', 'names', 'order', 'pal_park_encounters', 'pokedex_numbers', 'shape', 'varieties'])

In [16]:
links[0]['flavor_text_entries'][0]['flavor_text']

'A strange seed was\nplanted on its\nback at birth.\x0cThe plant sprouts\nand grows with\nthis POKéMON.'

In [17]:
text = []

for i in range(151):
    text.append(links[i]['flavor_text_entries'][0]['flavor_text'])
text

['A strange seed was\nplanted on its\nback at birth.\x0cThe plant sprouts\nand grows with\nthis POKéMON.',
 'When the bulb on\nits back grows\nlarge, it appears\x0cto lose the\nability to stand\non its hind legs.',
 'The plant blooms\nwhen it is\nabsorbing solar\x0cenergy. It stays\non the move to\nseek sunlight.',
 'Obviously prefers\nhot places. When\nit rains, steam\x0cis said to spout\nfrom the tip of\nits tail.',
 'When it swings\nits burning tail,\nit elevates the\x0ctemperature to\nunbearably high\nlevels.',
 'Spits fire that\nis hot enough to\nmelt boulders.\x0cKnown to cause\nforest fires\nunintentionally.',
 'Shoots water at\nprey while in the\nwater.\x0cWithdraws into\nits shell when in\ndanger.',
 'When tapped, this\nPOKéMON will pull\nin its head, but\x0cits tail will\nstill stick out a\nlittle bit.',
 'It deliberately\nmakes itself heavy\nso it can with\xad\x0cstand the recoil\nof the water jets\nit fires.',
 'It releases a stench from its red\nantenna to repel enemies. I

Let's clean some weird characters and add them to our current dataframe

In [18]:
info = [t.replace('\n', ' ').replace('\x0c',' ').replace('\u3000',' ').replace('\xad ', '') for t in text]

In [19]:
info

['A strange seed was planted on its back at birth. The plant sprouts and grows with this POKéMON.',
 'When the bulb on its back grows large, it appears to lose the ability to stand on its hind legs.',
 'The plant blooms when it is absorbing solar energy. It stays on the move to seek sunlight.',
 'Obviously prefers hot places. When it rains, steam is said to spout from the tip of its tail.',
 'When it swings its burning tail, it elevates the temperature to unbearably high levels.',
 'Spits fire that is hot enough to melt boulders. Known to cause forest fires unintentionally.',
 'Shoots water at prey while in the water. Withdraws into its shell when in danger.',
 'When tapped, this POKéMON will pull in its head, but its tail will still stick out a little bit.',
 'It deliberately makes itself heavy so it can withstand the recoil of the water jets it fires.',
 'It releases a stench from its red antenna to repel enemies. It grows by molting repeatedly.',
 'This POKéMON is vulnerable to atta

In [20]:
backup = df.copy()

In [21]:
df['information'] = info

In [22]:
df.head(5)

number        name type_1  type_2  total  hp  attack  defense  sp_atk  \
0       1   Bulbasaur  Grass  Poison    318  45      49       49      65   
1       2     Ivysaur  Grass  Poison    405  60      62       63      80   
2       3    Venusaur  Grass  Poison    525  80      82       83     100   
3       4  Charmander   Fire     NaN    309  39      52       43      60   
4       5  Charmeleon   Fire     NaN    405  58      64       58      80   

   sp_def  speed  islegendary  color  hasgender  height_m  weight_kg  \
0      65     45        False  Green       True      0.71        6.9   
1      80     60        False  Green       True      0.99       13.0   
2     100     80        False  Green       True      2.01      100.0   
3      50     65        False    Red       True      0.61        8.5   
4      65     80        False    Red       True      1.09       19.0   

   catch_rate      body_style  \
0          45       quadruped   
1          45       quadruped   
2          45       quadruped   
3          45  bipedal_tailed   
4          45  bipedal_tailed   

                                         information  
0  A strange seed was planted on its back at birt...  
1  When the bulb on its back grows large, it appe...  
2  The plant blooms when it is absorbing solar en...  
3  Obviously prefers hot places. When it rains, s...  
4  When it swings its burning tail, it elevates t...

Done! Let us add the apis per Pokémon as well:

In [23]:
BASE_URL = 'https://pokeapi.co/api/v2/'
midpoint = 'pokemon/'
apis = []

for i in range(152):
    if i != 0:
        api = BASE_URL + midpoint + str(i)
        apis.append(api)
apis

['https://pokeapi.co/api/v2/pokemon/1',
 'https://pokeapi.co/api/v2/pokemon/2',
 'https://pokeapi.co/api/v2/pokemon/3',
 'https://pokeapi.co/api/v2/pokemon/4',
 'https://pokeapi.co/api/v2/pokemon/5',
 'https://pokeapi.co/api/v2/pokemon/6',
 'https://pokeapi.co/api/v2/pokemon/7',
 'https://pokeapi.co/api/v2/pokemon/8',
 'https://pokeapi.co/api/v2/pokemon/9',
 'https://pokeapi.co/api/v2/pokemon/10',
 'https://pokeapi.co/api/v2/pokemon/11',
 'https://pokeapi.co/api/v2/pokemon/12',
 'https://pokeapi.co/api/v2/pokemon/13',
 'https://pokeapi.co/api/v2/pokemon/14',
 'https://pokeapi.co/api/v2/pokemon/15',
 'https://pokeapi.co/api/v2/pokemon/16',
 'https://pokeapi.co/api/v2/pokemon/17',
 'https://pokeapi.co/api/v2/pokemon/18',
 'https://pokeapi.co/api/v2/pokemon/19',
 'https://pokeapi.co/api/v2/pokemon/20',
 'https://pokeapi.co/api/v2/pokemon/21',
 'https://pokeapi.co/api/v2/pokemon/22',
 'https://pokeapi.co/api/v2/pokemon/23',
 'https://pokeapi.co/api/v2/pokemon/24',
 'https://pokeapi.co/api/

In [24]:
df['api'] = apis

In [25]:
df.head()

number        name type_1  type_2  total  hp  attack  defense  sp_atk  \
0       1   Bulbasaur  Grass  Poison    318  45      49       49      65   
1       2     Ivysaur  Grass  Poison    405  60      62       63      80   
2       3    Venusaur  Grass  Poison    525  80      82       83     100   
3       4  Charmander   Fire     NaN    309  39      52       43      60   
4       5  Charmeleon   Fire     NaN    405  58      64       58      80   

   sp_def  speed  islegendary  color  hasgender  height_m  weight_kg  \
0      65     45        False  Green       True      0.71        6.9   
1      80     60        False  Green       True      0.99       13.0   
2     100     80        False  Green       True      2.01      100.0   
3      50     65        False    Red       True      0.61        8.5   
4      65     80        False    Red       True      1.09       19.0   

   catch_rate      body_style  \
0          45       quadruped   
1          45       quadruped   
2          45       quadruped   
3          45  bipedal_tailed   
4          45  bipedal_tailed   

                                         information  \
0  A strange seed was planted on its back at birt...   
1  When the bulb on its back grows large, it appe...   
2  The plant blooms when it is absorbing solar en...   
3  Obviously prefers hot places. When it rains, s...   
4  When it swings its burning tail, it elevates t...   

                                   api  
0  https://pokeapi.co/api/v2/pokemon/1  
1  https://pokeapi.co/api/v2/pokemon/2  
2  https://pokeapi.co/api/v2/pokemon/3  
3  https://pokeapi.co/api/v2/pokemon/4  
4  https://pokeapi.co/api/v2/pokemon/5

Success! Let's proceed with some web scraping

# Web Scraping using Selenium

In [26]:
!pip install selenium

You should consider upgrading via the 'c:\users\ffrei\appdata\local\programs\python\python39\python.exe -m pip install --upgrade pip' command.


In [27]:
from selenium import webdriver
import requests

In [28]:
PATH='chromedriver.exe'

We will extract information from wikipedia. Thus, one should consider the beloow url:

In [29]:
url = 'https://en.wikipedia.org/wiki/List_of_generation_I_Pok%C3%A9mon'

In [30]:
driver=webdriver.Chrome(PATH)
driver.get(url)

In [31]:
table=driver.find_element_by_xpath('//*[@id="mw-content-text"]/div[1]/table[2]')
table=table.find_element_by_tag_name('tbody')
rows=table.find_elements_by_tag_name('tr')
elements=[[e.text for e in r.find_elements_by_tag_name('td')] for r in rows]
elements

[['1',
  'Grass',
  'Poison',
  'Ivysaur (#002)',
  'Reception to Bulbasaur has been largely positive and it often appears in "top Pokémon lists".[3][4] Its English name is a portmanteau of "bulb" and "dinosaur".[5] It shares its category with Ivysaur, Venusaur and Sunkern. They are all known as the Seed Pokémon.'],
 ['2',
  'Grass',
  'Poison',
  'Venusaur (#003)',
  'Playable character in Super Smash Bros. Brawl and Super Smash Bros. Ultimate as a part of the "Pokémon Trainer" fighter. Its English name is a portmanteau of "ivy" and "dinosaur".[6]'],
 ['3',
  'Grass',
  'Poison',
  'Mega Evolution',
  '[Gigantamax] Its English name is a portmanteau of "Venus" (relating to the Venus flytrap) and "dinosaur". It is the mascot of Pokémon Green and LeafGreen.[7] It is based on the Pareiasaur.'],
 ['4',
  'Fire',
  'Charmeleon (#005)',
  'Charmander is a bipedal, salamander-like creature with a flame at the tip of its tail. Its English name is a portmanteau of "char" and "salamander".[8]'],

In [32]:
elements[5][4]

'[Gigantamax] Playable character as a part of the "Pokémon Trainer" fighter in Super Smash Bros. Brawl and Super Smash Bros. Ultimate, as well as a standalone fighter in Super Smash Bros. for Nintendo 3DS and Wii U. Its English name is a portmanteau of "char" and "lizard".[10] Its Gigantamax Form has literal fire for wings and has fire coming out of its body. Mascot of Pokémon Red and FireRed.'

We need to make a workaround to collect the notes information column. It happens that our table has some merged cells either on the Type(s) heads or two different Pokémon share the same note. Therefore, the lenght of our elements list above can either be:

   + 5: in that case, we should collect the information within elements [i][4]
   + 4: in that case, we should collect the information within elements [i][3]
   + 3: in that case, we should collect the information within elements [i][2]
   
Note that the same information applies for any other information we might want to collect from the dataset. This is just a particular case. In general case would be:

   + 5: in that case, we should collect the information within elements [i][n]
   + 4: in that case, we should collect the information within elements [i][n-1]
   + 3: in that case, we should collect the information within elements [i][n-2]

Being n the index for a given position vector within the maximum indexes of the matrix elements

In [33]:
lenght=[]
for e in elements:
    lenght.append(len(e))
    
set(lenght)

{3, 4, 5}

In [34]:
count=0
for e, l in enumerate(lenght):
    if l == 3:
        count += 1
        print(e)
count

100
116
118


3

There are 3 vectors/ 4 Pokémon that are missing this note information they are (print(e) allows us to identify them):

   + #100 Voltorb
   + #101 Electrode
   + #116 Horsea
   + #117 Seadra
   
Let's collect the notes, taken into account that the information within the mentioned column either doesn't exist at all or exists in one time in only the above mentioned

In [35]:
notes = []
evolutions = []

for i, e in enumerate(elements):
    if len(e) == 5:
        notes.append(elements[i][4])
        evolutions.append(elements[i][3])
    elif len(e) == 4:
        notes.append(elements[i][3])
        evolutions.append(elements[i][2])
    else:
        notes.append(elements[i][2])
        evolutions.append(elements[i][1])
        
len(notes) == len(evolutions)


True

In [36]:
notes

['Reception to Bulbasaur has been largely positive and it often appears in "top Pokémon lists".[3][4] Its English name is a portmanteau of "bulb" and "dinosaur".[5] It shares its category with Ivysaur, Venusaur and Sunkern. They are all known as the Seed Pokémon.',
 'Playable character in Super Smash Bros. Brawl and Super Smash Bros. Ultimate as a part of the "Pokémon Trainer" fighter. Its English name is a portmanteau of "ivy" and "dinosaur".[6]',
 '[Gigantamax] Its English name is a portmanteau of "Venus" (relating to the Venus flytrap) and "dinosaur". It is the mascot of Pokémon Green and LeafGreen.[7] It is based on the Pareiasaur.',
 'Charmander is a bipedal, salamander-like creature with a flame at the tip of its tail. Its English name is a portmanteau of "char" and "salamander".[8]',
 'Charmeleon shares a nearly identical design to Charmander, though is larger in stature. Its English name is a portmanteau of "char" and "chameleon".[9]',
 '[Gigantamax] Playable character as a par

The column contains Missing no which an error and a wholer other topic. Let's remove it for now!

In [37]:
note = notes[:151]
evol =  evolutions[:151]
evol

['Ivysaur (#002)',
 'Venusaur (#003)',
 'Mega Evolution',
 'Charmeleon (#005)',
 'Charizard (#006)',
 'Two Mega Evolutions',
 'Wartortle (#008)',
 'Blastoise (#009)',
 'Mega Evolution',
 'Metapod (#011)',
 'Butterfree (#012)',
 'End of evolution',
 'Kakuna (#014)',
 'Beedrill (#015)',
 'Mega Evolution',
 'Pidgeotto (#017)',
 'Pidgeot (#018)',
 'Mega Evolution',
 'Raticate (#020)',
 'End of evolution',
 'Fearow (#022)',
 'End of evolution',
 'Arbok (#024)',
 'End of evolution',
 'Raichu (#026)',
 'End of evolution',
 'Sandslash (#028)',
 'End of evolution',
 'Nidorina (#030)',
 'Nidoqueen (#031)',
 'End of evolution',
 'Nidorino (#033)',
 'Nidoking (#034)',
 'End of evolution',
 'Clefable (#036)',
 'End of evolution',
 'Ninetales (#038)',
 'End of evolution',
 'Wigglytuff (#040)',
 'End of evolution',
 'Golbat (#042)',
 'Crobat (#169)',
 'Gloom (#044)',
 'Vileplume (#045)\nBellossom (#182)',
 'End of evolution',
 'Parasect (#047)',
 'End of evolution',
 'Venomoth (#049)',
 'End of evolu

In [38]:
evolution = [ev.replace('[nb 7]', '',)\
             .replace('\nSlowking (#199)','')\
             .replace('\nPolitoed (#186)','')\
             .replace('\nBellossom (#182)','')\
             .replace('Crobat (#169)','End of evolution')\
             .replace('Mega Evolution','End of evolution')\
             .replace('Rhyperior (#464)','End of evolution')\
             .replace('Blissey (#242)','End of evolution')\
             .replace('Two End of evolutions','End of evolution')
             .replace('Magmortar (#467)','End of evolution')
             .replace('Electivire (#466)','End of evolution')
             .replace('Scizor (#212)','End of evolution')
             .replace('[nb 6]','End of evolution')
             .replace('Water','End of evolution')
             .replace('Tangrowth (#465)','End of evolution')
             for ev in evol]

In [39]:
evolution

['Ivysaur (#002)',
 'Venusaur (#003)',
 'End of evolution',
 'Charmeleon (#005)',
 'Charizard (#006)',
 'End of evolution',
 'Wartortle (#008)',
 'Blastoise (#009)',
 'End of evolution',
 'Metapod (#011)',
 'Butterfree (#012)',
 'End of evolution',
 'Kakuna (#014)',
 'Beedrill (#015)',
 'End of evolution',
 'Pidgeotto (#017)',
 'Pidgeot (#018)',
 'End of evolution',
 'Raticate (#020)',
 'End of evolution',
 'Fearow (#022)',
 'End of evolution',
 'Arbok (#024)',
 'End of evolution',
 'Raichu (#026)',
 'End of evolution',
 'Sandslash (#028)',
 'End of evolution',
 'Nidorina (#030)',
 'Nidoqueen (#031)',
 'End of evolution',
 'Nidorino (#033)',
 'Nidoking (#034)',
 'End of evolution',
 'Clefable (#036)',
 'End of evolution',
 'Ninetales (#038)',
 'End of evolution',
 'Wigglytuff (#040)',
 'End of evolution',
 'Golbat (#042)',
 'End of evolution',
 'Gloom (#044)',
 'Vileplume (#045)',
 'End of evolution',
 'Parasect (#047)',
 'End of evolution',
 'Venomoth (#049)',
 'End of evolution',
 'D

Let's move on and add these columns to our DataFrame and proceed to load it to SQL

In [40]:
df['evolution'] = evolution
df['note'] = note

In [41]:
df.tail()

number       name   type_1  type_2  total   hp  attack  defense  sp_atk  \
146     147    Dratini   Dragon     NaN    300   41      64       45      50   
147     148  Dragonair   Dragon     NaN    420   61      84       65      70   
148     149  Dragonite   Dragon  Flying    600   91     134       95     100   
149     150     Mewtwo  Psychic     NaN    680  106     110       90     154   
150     151        Mew  Psychic     NaN    600  100     100      100     100   

     sp_def  ...   color  hasgender height_m  weight_kg  catch_rate  \
146      50  ...    Blue       True     1.80        3.3          45   
147      70  ...    Blue       True     3.99       16.5          45   
148     100  ...   Brown       True     2.21      210.0          45   
149      90  ...  Purple      False     2.01      122.0           3   
150     100  ...    Pink      False     0.41        4.0          45   

          body_style                                        information  \
146  serpentine_body  Long considered a mythical POKéMON until recen...   
147  serpentine_body  A mystical POKéMON that exudes a gentle aura. ...   
148   bipedal_tailed  An extremely rarely seen marine POKéMON. Its i...   
149   bipedal_tailed  It was created by a scientist after years of h...   
150   bipedal_tailed  So rare that it is still said to be a mirage b...   

                                       api         evolution  \
146  https://pokeapi.co/api/v2/pokemon/147  Dragonair (#148)   
147  https://pokeapi.co/api/v2/pokemon/148  Dragonite (#149)   
148  https://pokeapi.co/api/v2/pokemon/149  End of evolution   
149  https://pokeapi.co/api/v2/pokemon/150  End of evolution   
150  https://pokeapi.co/api/v2/pokemon/151      No evolution   

                                                  note  
146  Its existence was unknown until a fisherman pu...  
147                                                     
148  Pseudo-Legendary. It zooms off over seas to ch...  
149  Legendary. One of the playable characters in S...  
150  Only available through Nintendo events. It can...  

[5 rows x 22 columns]

Finally, let's give an obvious name to our DataFrame:

In [42]:
pokedex = df

backup = pokedex.copy()

In [43]:
pokedex.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 151 entries, 0 to 150
Data columns (total 22 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   number       151 non-null    int64  
 1   name         151 non-null    object 
 2   type_1       151 non-null    object 
 3   type_2       67 non-null     object 
 4   total        151 non-null    int64  
 5   hp           151 non-null    int64  
 6   attack       151 non-null    int64  
 7   defense      151 non-null    int64  
 8   sp_atk       151 non-null    int64  
 9   sp_def       151 non-null    int64  
 10  speed        151 non-null    int64  
 11  islegendary  151 non-null    bool   
 12  color        151 non-null    object 
 13  hasgender    151 non-null    bool   
 14  height_m     151 non-null    float64
 15  weight_kg    151 non-null    float64
 16  catch_rate   151 non-null    int64  
 17  body_style   151 non-null    object 
 18  information  151 non-null    object 
 19  api     

In [44]:
pokedex['type_2'].fillna('no type', inplace = True)

# Loading into SQL

In [45]:
!pip install mysql-connector-python

You should consider upgrading via the 'c:\users\ffrei\appdata\local\programs\python\python39\python.exe -m pip install --upgrade pip' command.


In [46]:
import mysql.connector as conn

**DON'T RUN THE BELOW CODE**

def create_database(name, user = 'root', password = '*****', check=True):
    
    create_database = conn.connect(host='localhost', user=user, passwd=password)
    
    cursor = create_database.cursor()
    
    cursor.execute(f'drop database if exists {name}')
    cursor.execute(f'create database {name}')
    
    print('Data Base created. ;)')
    
    if check:
        cursor.execute('show databases')
        for x in cursor:
            print(x)

In [ ]:
create_database('pokemon')

def create_table(df, table_name, data_base, user='root', password='*****', check=True, drop=False):
    
    db=conn.connect(host='localhost', user=user, passwd=password, database=data_base)
    cursor=db.cursor()
    
    if drop:
        cursor.execute(f'drop table if exists {table_name}')
    
    table = f"create table {table_name}(id int, {' varchar(255),'.join(df.columns)} varchar(2000));"
    print(table)
    
    cursor.execute(table)
    
    print(f'Done, table {table_name} created. ;)')
        
    if check:
        cursor.execute('show tables')
        for x in cursor:
            print(x)

In [50]:
create_table(pokedex, '1gen' , 'pokemon')

create table 1gen(id int, number varchar(255),name varchar(255),type_1 varchar(255),type_2 varchar(255),total varchar(255),hp varchar(255),attack varchar(255),defense varchar(255),sp_atk varchar(255),sp_def varchar(255),speed varchar(255),islegendary varchar(255),color varchar(255),hasgender varchar(255),height_m varchar(255),weight_kg varchar(255),catch_rate varchar(255),body_style varchar(255),information varchar(255),api varchar(255),evolution varchar(255),note varchar(2000));
Done, table 1gen created. ;)
('1gen',)


def insert_data(df, table_name, data_base, user='root', password= '*****'):
    
    db = conn.connect(host='localhost', user=user, passwd=password, database=data_base)
    cursor = db.cursor()
    
    for i in range(len(df)):
        
        insert_query = 'insert into {} (id, {}) values {};'\
                     .format(table_name, ','.join(df.columns), tuple(np.insert(df.iloc[i].values, 0, i)))
        
        cursor.execute(insert_query)
    
    db.commit()
    print(f'Done, data inserted into table {table_name}. ;)')
    print()
    print(insert_query)

In [52]:
insert_data(pokedex, '1gen' , 'pokemon')

Done, data inserted into table 1gen. ;)

insert into 1gen (id, number,name,type_1,type_2,total,hp,attack,defense,sp_atk,sp_def,speed,islegendary,color,hasgender,height_m,weight_kg,catch_rate,body_style,information,api,evolution,note) values (150, 151, 'Mew', 'Psychic', 'no type', 600, 100, 100, 100, 100, 100, 100, False, 'Pink', False, 0.41, 4.0, 45, 'bipedal_tailed', 'So rare that it is still said to be a mirage by many experts. Only a few people have seen it worldwide.', 'https://pokeapi.co/api/v2/pokemon/151', 'No evolution', 'Only available through Nintendo events. It can turn invisible at will and it can turn into anything at will.');


def read_sql(query, table_name, data_base, user='root', password='*****'):
    
    db = conn.connect(host='localhost', user=user, passwd=password, database=data_base)
    
    data = pd.read_sql(query, db)
    
    return data

In [61]:
read_sql('select * from 1gen limit 5', '1gen', 'pokemon')

id number        name type_1   type_2 total  hp attack defense sp_atk  ...  \
0   0      1   Bulbasaur  Grass   Poison   318  45     49      49     65  ...   
1   1      2     Ivysaur  Grass   Poison   405  60     62      63     80  ...   
2   2      3    Venusaur  Grass   Poison   525  80     82      83    100  ...   
3   3      4  Charmander   Fire  no type   309  39     52      43     60  ...   
4   4      5  Charmeleon   Fire  no type   405  58     64      58     80  ...   

   color hasgender height_m weight_kg catch_rate      body_style  \
0  Green         1     0.71       6.9         45       quadruped   
1  Green         1     0.99      13.0         45       quadruped   
2  Green         1     2.01     100.0         45       quadruped   
3    Red         1     0.61       8.5         45  bipedal_tailed   
4    Red         1     1.09      19.0         45  bipedal_tailed   

                                         information  \
0  A strange seed was planted on its back at birt...   
1  When the bulb on its back grows large, it appe...   
2  The plant blooms when it is absorbing solar en...   
3  Obviously prefers hot places. When it rains, s...   
4  When it swings its burning tail, it elevates t...   

                                   api          evolution  \
0  https://pokeapi.co/api/v2/pokemon/1     Ivysaur (#002)   
1  https://pokeapi.co/api/v2/pokemon/2    Venusaur (#003)   
2  https://pokeapi.co/api/v2/pokemon/3   End of evolution   
3  https://pokeapi.co/api/v2/pokemon/4  Charmeleon (#005)   
4  https://pokeapi.co/api/v2/pokemon/5   Charizard (#006)   

                                                note  
0  Reception to Bulbasaur has been largely positi...  
1  Playable character in Super Smash Bros. Brawl ...  
2  [Gigantamax] Its English name is a portmanteau...  
3  Charmander is a bipedal, salamander-like creat...  
4  Charmeleon shares a nearly identical design to...  

[5 rows x 23 columns]